### PR TITLE
Move Spark backend to backends/ directory

### LINKED ISSUE #2416: Move SPARK backend to the backends/ directory
**Move the `spark/` backend to `backends/spark`.**

The current Ibis directory structure is a bit confusing, since backends are mixed with core modules. For example, it's not obvious to know if `pandas/` is a backend, or core functionality. Many of the backends live in the root directory, and some are grouped, like in a directory `sql/`. This makes things even harder and difficult to maintain. Also, ibis uses attributes to access the backends (e.g. `ibis.spark.connect`), and `ibis.spark` in code seems to be the `ibis/spark/` module, but it's not, making things confusing, and even creating problems with pytest.

We're moving backends to a standardized structure, in the `backends/` directory, one backend at a time (each backend in a separate pull request). We started by the OmniSciDB backend, and its PR can be a reference for the changes that need to be made here: #2392

The exact tasks to perform are:
- Move the backend from its original location to `backends/` (if the backend is in a subdirectory, like `sql/`, ignore the subdirectory and just have the backend inside `backends/`, so `sql/sqlite/` -> `backends/sqlite`)
- Move the content of `backends/<your-backend>/api.py` to `backends/<your-backend>/__init__.py`. Backends are using `api.py` for historical reasons, but the public backend API should be in the standard `__init__.py` file of the module.
- Update `ibis/__init__.py` so the backend is imported from the new module (`from ibis.<your-backend> import api` -> `from ibis import <your-backend>`)
- Look for places in the code where your backend is being imported, and update the imports. For imports inside the same backend, please use relative imports (e.g. `import ibis.<your-backend>.whatever` -> `import .whatever`). See the reference PR for examples.
- Update the documentation (in `doc/source/backends/<your-backend>.rst`) so the API is generated from the right location.

After performing the changes, the CI should be green, except for the packaging build (which requires changes in the conda-forge feedstock). You can ignore that error, those changes will be performed later for all moved backends.

### PR DESCRIPTION
**Pull request by EdAbati**
As part of the Ibis sprint, I am making this PR that closes #2416

Please let me know if if and how I can improve it.

Thanks,

### CODE REVIEW

File: ibis/__init__.py [hunk 1]
```diff
@@ -57,7 +57,7 @@
 
 with suppress(ImportError):
     # pip install ibis-framework[spark]
-    import ibis.spark.api as spark  # noqa: F401
+    import ibis.backends as spark  # noqa: F401
```
Comment (datapythonista) [inline ibis/__init__.py:60]:
```suggestion
    from ibis.backends import spark  # noqa: F401
```

File: ibis/backends/spark/tests/test_ddl_compilation.py [hunk 2]
```diff
@@ -2,11 +2,11 @@
 
 import ibis
 
-from ibis.spark.compiler import (  # noqa: E402, isort:skip
+from ibis.backends.spark.compiler import (  # noqa: E402, isort:skip
```
Comment (datapythonista) [inline ibis/backends/spark/tests/test_ddl_compilation.py:5]:
```suggestion
from .compiler import (  # noqa: E402, isort:skip
```

File: ibis/backends/spark/tests/test_ddl_compilation.py [hunk 3]
```diff
@@ -2,11 +2,11 @@
...
     SparkDialect,
     build_ast,
 )
-from ibis.spark import ddl  # noqa: E402, isort:skip
+from ibis.backends.spark import ddl  # noqa: E402, isort:skip
```
Comment (datapythonista) [inline ibis/backends/spark/tests/test_ddl_compilation.py:9]:
```suggestion
from . import ddl  # noqa: E402, isort:skip
```

File: ibis/backends/spark/tests/test_udf.py [hunk 4]
```diff
@@ -13,7 +13,7 @@
 
 py4j = pytest.importorskip('py4j')
 ps = pytest.importorskip('pyspark')
-from ibis.spark.udf import udf  # noqa: E402, isort:skip
+from ibis.backends.spark.udf import udf  # noqa: E402, isort:skip
```
Comment (datapythonista) [inline ibis/backends/spark/tests/test_udf.py:16]:
```suggestion
from .udf import udf  # noqa: E402, isort:skip
```

File: ibis/backends/spark/tests/test_udf.py [hunk 5]
```diff
@@ -147,7 +147,7 @@ def quantiles(series, quantiles):
 
 
 def test_spark_dtype_to_ibis_dtype():
-    from ibis.spark.datatypes import _SPARK_DTYPE_TO_IBIS_DTYPE
+    from ibis.backends.spark import _SPARK_DTYPE_TO_IBIS_DTYPE
```
Comment (datapythonista) [inline ibis/backends/spark/tests/test_udf.py:150]:
```suggestion
    from . import _SPARK_DTYPE_TO_IBIS_DTYPE
```

**Review by datapythonista — COMMENTED** [review]
Thanks @EdAbati, nice work. Just few changes in the imports, but looks good.

**Review by icexelloss — CHANGES_REQUESTED** [review]
Please do not merge. Let me investigate CI.

**Comment by jreback** [comment]
lgtm, conflict.

**Comment by datapythonista** [comment]
@icexelloss can we merge this already? I guess we can be confident about the tests after #2442, right?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @EdAbati

**Pull request merged**